### PR TITLE
Keep at most one empty line at the end of each example

### PR DESCRIPTION
**Pull request by krlmlr**
This is crucial for combining multiple `@examples` and `@examplesIf` elements.

### CODE REVIEW

**Comment by lorenzwalthert** [comment]
I am not sure I can follow. Can you provide a new test to demonstrate the expected behavior?

**Comment by lorenzwalthert** [comment]
Also, we don't squash examples anymore: https://github.com/r-lib/styler/issues/747

**Comment by krlmlr** [comment]
Thanks. The "examples" are tibble and pillar -- try styling e.g. `add.R` in tibble with this PR and with the main branch. I'll add a formal test tomorrow.

**Comment by lorenzwalthert** [comment]
Ok, so what's the rationale of having a blank line like between roxygen comments this?


```r

#' Add columns to a data frame
#'
#' This is a convenient way to add one or more columns to an existing data
#' frame.
#'
#' @param .data Data frame to append to.
#' @param ... <[`dynamic-dots`][rlang::dyn-dots]>
#'   Name-value pairs, passed on to [tibble()]. All values must have
#'   the same size of `.data` or size 1.
#' @param .before,.after One-based column index or column name where to add the
#'   new columns, default: after last column.
#' @inheritParams tibble
#' @family addition
#' @examples
#' # add_column ---------------------------------
#' df <- tibble(x = 1:3, y = 3:1)
#'
#' df %>% add_column(z = -1:1, w = 0)
#' df %>% add_column(z = -1:1, .before = "y")
#'
#' # You can't overwrite existing columns
#' try(df %>% add_column(x = 4:6))
#'

#' # You can't create new observations
#' try(df %>% add_column(z = 1:5))
#'
#' @export
add_column
```

**Comment by lorenzwalthert** [comment]
having blocks in the docs like here?
Screenshot 2021-12-22 at 18 55 47

**Comment by krlmlr** [comment]
No, the blank line is bogus and can be removed. This is about the empty line before `#' @export`, the help page contains other functions with more examples, and I need to keep that empty line to visually separate the examples from each other.

**Comment by krlmlr** [comment]
I have updated the test output files, there's still trouble. Could you please take a look?

Also, there are "unknown or unitialised token column" warnings, I have seen them before with CRAN styler.

I do wonder why this change adds so many empty lines for some outputs, in particular in the presence of `\dontrun{}`.

**Comment by krlmlr** [comment]
Empty lines look good to me now. Would you support this change?

**Comment by krlmlr** [comment]
Regarding the tidyverse style guide: not sure, it just feels that examples would be too condensed without empty lines inbetween.

**Comment by lorenzwalthert** [comment]
If I understand you correctly from your last two comments, all we need is to keep an empty line at the end of an example section **if** already one or more empty lines are present (empty as in only `#'` on that line, not completely empty). If there is none, we should not add one. Correct?

**Comment by krlmlr** [comment]
Yes, that would be great.


On 23 Dec 2021 at 12:51:14 CET, Lorenz Walthert ***@***.***> wrote:

   If I understand you correctly from your last two comments, all we need is to keep an empty line at the end of an example section  if  already one or more empty lines are present (empty as in only  #'  on that line, not completely empty). If there is none, we should not add one. Correct?  —
Reply to this email directly,  view it on GitHub , or  unsubscribe .
Triage notifications on the go with GitHub Mobile for  iOS  or  Android .

**Comment by lorenzwalthert** [comment]
The warning about unknown or uninitialized column seems to happen in various places in the code (as you can tell from the warnings). And it's not new. Probably it's similar to this:

``` r
x <- tibble::tibble()
x$ff[logical()] <- 1
#> Warning: Unknown or uninitialised column: `ff`.
```

<sup>Created on 2021-12-23 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

For some reason, `pd` has sometimes now rows and the condition for the assignment is `logical()`, which has length 0.

Also, don't modify the `.*-utf8` reference objects. I don't know why they don't fail on CRAN, but they fail for me locally too. Wird uns wohl irgendwann um die Ohren fliegen 😄

**Comment by lorenzwalthert** [comment]
@krlmlr can you have a look at the new test case I added and if that's what you want, you can merge it 🥳

File: NEWS.md [hunk 1]
```diff
@@ -37,6 +37,7 @@
   detected (#881).
 * Add vignette on distributing style guide (#846, #861).
 * ensure a trailing blank line also if the input is cached (#867).
+* Preserve trailing blank line in roxygen examples (#880).
```
Comment (krlmlr) [inline NEWS.md:40]:
```suggestion
* Preserve trailing blank line in roxygen examples to simplify concatenation of examples (#880).
```

**Comment by krlmlr** [comment]
Are you suggesting that `pd` has no columns sometimes? Even if it has no rows and the column exists, the assignment works:

``` r
x <- tibble::tibble(ff = numeric())
x$ff[logical()] <- 1
x
#> # A tibble: 0 × 1
#> # … with 1 variable: ff <dbl>
```

<sup>Created on 2021-12-24 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

We should normalize the data early in the process so that the structure is stable.

**Comment by lorenzwalthert** [comment]
> Are you suggesting that pd has no columns sometimes? 

Well, it has columns. But I think the problem is that it does not habe the ones to which something is assigned. Comparing my example with yours, yours already has the column that is assigned to. I’ll open a nee issue.

**Pull request merged**